framework/source/class/qx/util/DynamicScriptLoader.js

```diff
@@ -0,0 +1,234 @@
+     */
+    ready: 'qx.event.type.Event'
+  },
+
+  statics: {
+    /**
+     * Map of scripts being added at the present time. Key is script name; value is instance of this class which
+     * is loading it.
+     */
+    __IN_PROGRESS: {},
+     /**
+      * Map of scripts that have fully loaded. Key is script name; value is true
+      */
+    __LOADED: {}
+  },
+
+  /*
+  *****************************************************************************
+     MEMBERS
+  *****************************************************************************
+  */
+
+
+  members: {
+
+    /**
+     * Array of the scripts to be loaded
+     */
+    __QUEUE: null,
+
+    /**
+     * Start loading scripts. This may only be called once!
+     *
+     */
+    start: function() {
+      this.start = function(){
+        throw new Error(this.tr('you can only call start once per instance'));
+      }
+      this.__loadScripts();
+    },
+
+
+    /**
+     * Chain loading scripts.
+     *
+     * Recursively called until the array of scripts is consumed
+     *
+     */
+    __loadScripts: function() {
+      var cl = qx.util.DynamicScriptLoader;
+      var script;
+      var dynLoader;
+      var id1, id2;
+      var uri;
+      var loader;
+
+      script = this.__QUEUE.shift();
+      if (!script){
+        this.fireEvent("ready")
+        return;
+      }
+
+      if (cl.__LOADED[script]){
+        this.fireDataEvent('loaded',{
+          script: script,
+          status: 'preloaded'
+        });
+        this.__loadScripts();
+        return;
+      }           
+
+      dynLoader = cl.__IN_PROGRESS[script];
+      if (dynLoader){
+
+          id1 = dynLoader.addListener('loaded',function(e){
+            var data = e.getData();
+            if (data.script === script){
+              dynLoader.removeListenerById(id2);
+              dynLoader.removeListenerById(id1);
+              this.fireDataEvent('loaded',data);
+              this.__loadScripts();
+            }
+          },this);
+
+          id2 = dynLoader.addListener('failed',function(e){
+            var data = e.getData();
+            dynLoader.removeListenerById(id1);
+            dynLoader.removeListenerById(id2);              
+            this.fireDataEvent('failed',{
+              script: script,
+              status: 'loading of ' + data.script + ' failed while waiting for ' + script
+            });
+          },this);
+
+          return;
+      }
+
+      uri = qx.util.ResourceManager.getInstance().toUri(script);
+
+      loader = new qx.bom.request.Script();
+
+      loader.on("load", function(request) {
+        cl.__LOADED[script] = true;
+        delete cl.__IN_PROGRESS[script];
+        this.fireDataEvent('loaded', {
+          script: script,
+          status: request.status
+        });
+        this.__loadScripts();
+      },this);
+
+      var onError = function(request) {
+        delete cl.__IN_PROGRESS[script];
```

Comment:
maybe add here ```cl.__LOADED[script] = false;``` to avoid trying to load the script next time again?

This would also need to distinguish the two states false and true in the pre-test if loading needs to be started:
```javascript
      if (cl.__LOADED[script] === true){
         this.fireDataEvent('loaded',{
           script: script,
           status: 'preloaded'
         });
         this.__loadScripts();
         return;
       }
       if (cl.__LOADED[script] === false){
         this.fireDataEvent('failed',{
           script: script,
           status: 'failed'
         });
         return;
       }
```